internal/clients/kubernetes/client.go

```diff
@@ -163,13 +170,74 @@ func makeClientKey(orgName, name string) string {
 }
 
 // GetK8sClient retrieves a Kubernetes client for the specified org and cluster.
+// Deprecated: Use GetK8sClientFromDataPlane instead
 func GetK8sClient(
 	clientMgr *KubeMultiClientManager,
 	orgName, name string,
 	kubernetesCluster openchoreov1alpha1.KubernetesClusterSpec,
 ) (client.Client, error) {
 	key := makeClientKey(orgName, name)
-	cl, err := clientMgr.GetClient(key, kubernetesCluster)
+	cl, err := clientMgr.GetClient(key, &kubernetesCluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Kubernetes client: %w", err)
+	}
+	return cl, nil
+}
+
+// GetK8sClientFromDataPlane retrieves a Kubernetes client from DataPlane specification.
+// It automatically handles agent mode vs direct access mode.
+func GetK8sClientFromDataPlane(
+	clientMgr *KubeMultiClientManager,
+	dataplane *openchoreov1alpha1.DataPlane,
+	agentServer interface{}, // *server.Server, passed as interface to avoid circular dependency
+) (client.Client, error) {
+	key := makeClientKey(dataplane.Namespace, dataplane.Name)
+
+	if dataplane.Spec.Agent != nil && dataplane.Spec.Agent.Enabled {
+		// Agent mode - check if we already have a client
+		clientMgr.mu.Lock()
+		if cl, exists := clientMgr.clients[key]; exists {
+			clientMgr.mu.Unlock()
+			return cl, nil
+		}
+		clientMgr.mu.Unlock()
+
+		if agentServer == nil {
+			return nil, fmt.Errorf("agent server is required for agent mode but not provided")
+		}
+
+		srv, ok := agentServer.(agentserver.Dispatcher)
+		if !ok {
+			return nil, fmt.Errorf("invalid agent server type: expected agentserver.Dispatcher")
+		}
+
+		// Get plane name for agent routing
+		planeName := dataplane.Name
+		if dataplane.Spec.Agent.ClusterName != "" {
+			planeName = dataplane.Spec.Agent.ClusterName
+		}
+
+		// Create agent client with 30 second timeout for requests
+		requestTimeout := 30 * time.Second
+		cl, err := NewAgentClient(planeName, srv, requestTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create agent client: %w", err)
+		}
+
+		// Cache the client
+		clientMgr.mu.Lock()
+		clientMgr.clients[key] = cl
+		clientMgr.mu.Unlock()
```

Comment:
encapsulate access to the map and locks - best not to expect clients to do locking for you